sycl/doc/extensions/proposed/sycl_ext_oneapi_graph.asciidoc

```diff
@@ -0,0 +1,1612 @@
+considered immediately stale in multi-threaded usage, as another thread could
+have preemptively changed the state of the queue.
+
+=== Exception Safety
+
+In addition to the destruction semantics provided by the SYCL
+{crs}[common reference semantics], when the last copy of a modifiable
+`command_graph` is destroyed recording is ended on any queues that are recording
+to that graph, equivalent to `+this->end_recording()+`.
+
+As a result, users don't need to manually wrap queue recording code in a
+`try` / `catch` block to reset the state of recording queues on an exception
+back to the executing state. Instead, an uncaught exception destroying the
+modifiable graph will perform this action, useful in RAII pattern usage.
+
+=== Error Handling
+
+When a queue is in recording mode asynchronous exceptions will not be
+generated, as no device execution is occurring. Synchronous errors specified as
+being thrown in the default queue executing state, will still be thrown when a
+queue is in the recording state. Queue query methods operate as usual in
+recording mode, as opposed to throwing.
+
+The `command_graph::begin_recording` and `command_graph::end_recording`
+entry-points return a `bool` value informing the user whether a related queue
+state change occurred. False is returned rather than throwing an exception when
+no queue state is changed. This design is because the queues are already in
+the state the user desires, so if the function threw an exception in this case,
+the application would likely swallow it and then proceed.
+
+=== Storage Lifetimes [[storage-lifetimes]]
+
+The lifetime of any buffer recorded as part of a submission
+to a command graph will be extended in keeping with the common reference
+semantics and buffer synchronization rules in the SYCL specification. It will be
+extended either for the lifetime of the graph (including both modifiable graphs
+and the executable graphs created from them) or until the buffer is no longer
+required by the graph (such as after being replaced through executable graph update).
+
+If a buffer created with a host data pointer is recorded as part of a submission to
+a command graph, the lifetime of that host data will also be extended by taking a
+copy of that data inside the buffer. This copy will occur during the call to
+`queue::submit()`. Users can opt-out of this behaviour by passing the property
+`graph::no_host_copy` to `command_graph::begin_recording()` or the constructor
+of the `command_graph` the commands will be recorded to. Passing the property to
+`begin_recording()` will prevent host copies only for commands recorded before
+`end_recording()` is called for a given queue. Passing the property to the
+`command_graph` constructor will prevent host copies for all commands recorded to
+the graph.
+
+The copy will not occur if the buffer is accessed inside the submitted command
+group function with an accessor with `access_mode::write` or the `no_init`
+property.
```

Comment:
For `access_mode::write` any data we copy can't be read by the command-group, so doing the copy isn't required for correctness.

With `no_init`, the user won't rely on the initial host data values, and the spec description of `no_init` says
"The implementation can use this information to avoid copying the accessor’s data in some cases".

Reply:
> For `access_mode::write` any data we copy can't be read by the command-group, so doing the copy isn't required for correctness.

Consider the case where the command writes only some of the entries.  In this case, the unwritten buffer entries must still have their original values.  How will you accomplish this if the host memory containing those value is no longer available?